### ISSUE TITLE
Inconsistent behaviour of IsomorphismFpSemigroup

### ISSUE THREAD

**Issue by mtorpey** [gap-system/gap#1258]
A mapping given by `IsomorphismFpSemigroup` and `InverseGeneralMapping` behaves as expected on the first run, but then throws an error when it is called again:
```
gap> F := FreeMonoid(2);;
gap> M := F / [[F.2 ^ 2, F.2]];;
gap> iso := IsomorphismFpSemigroup(M);;
gap> S := Range(iso);;
gap> x := S.2 * S.3 * S.2;;
gap> x ^ InverseGeneralMapping(iso);
m1*m2*m1
gap> x ^ InverseGeneralMapping(iso);
Error, Lists Assignment: <list> must be a mutable list in
  wlist[2 * i - 1] := wlist[2 * i - 1] - 1; at /home/mtorpey/gap/lib/fpmon.gi:434 called from 
FStoFM( One( FM ), UnderlyingElement( x ) ) at /home/mtorpey/gap/lib/fpmon.gi:466 called from
map!.fun( elm ) at /home/mtorpey/gap/lib/mapprep.gi:591 called from
ImageElm( map, elm ) at /home/mtorpey/gap/lib/mapping.gi:835 called from
<function "unknown">( <arguments> )
 called from read-eval loop at *stdin*:17
you can 'return;' and ignore the assignment
```

I'm not sure why `x ^ InverseGeneralMapping(iso);` works the first time, or why it fails the second time.  What's going on here?

### GAP banner
```
 ┌───────┐   GAP 4.8.6-3392-gea4ec96 of today
 │  GAP  │   https://www.gap-system.org
 └───────┘   Architecture: x86_64-pc-linux-gnu-gcc-default64
 Libs used:  gmp 6.1.2
 Loading the library and packages ...
 Components: trans 1.0, prim 3.0, small* 1.0, id* 1.0
 Packages:   GAPDoc 1.5.1, IO 4.4.6
 Try '??help' for help. See also '?copyright', '?cite' and '?authors'
```

**Comment by fingolfin** [comment]
Here is a patch which fixes the issue:
```diff
diff --git a/lib/fpmon.gi b/lib/fpmon.gi
index 7d5a0fbb8..b9c894f51 100644
--- a/lib/fpmon.gi
+++ b/lib/fpmon.gi
@@ -404,7 +404,7 @@ function(M)
   FMtoFS := function(id, w)
     local wlist, i;

-    wlist := ShallowCopy(ExtRepOfObj(w));
+    wlist := ExtRepOfObj(w);

     if Length(wlist) = 0 then # it is the identity
       return id;
@@ -412,6 +412,7 @@ function(M)

     # have to increment the generators by one to shift past the identity
     # generator
+    wlist := ShallowCopy(wlist);
     for i in [1 .. 1 / 2 * (Length(wlist))] do
       wlist[2 * i - 1] := wlist[2 * i - 1] + 1;
     od;
@@ -430,6 +431,7 @@ function(M)
     fi;

     # have to decrease each entry by one because of the identity generator
+    wlist := ShallowCopy(wlist);
     for i in [1 .. 1 / 2 * (Length(wlist))] do
       wlist[2 * i - 1] := wlist[2 * i - 1] - 1;
     od;
```

Note that the first part is not strictly needed, but it makes the code uniform. And doing the ShallowCopy after the identity check is very slightly more efficient, obviously.

If anybody wants to make a PR which fixes this, please feel free to, but make sure to include a test case which checks the code (e.g. based on the example code above)

**Comment by markuspf** [comment]
This is because the element type changes from `FreeSemigroupElementsFamily` to `FamilyElementsFpSemigroup`

**Comment by mtorpey** [comment]
This is fixed by PR #1259.
